servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ReplayStrategies.java

```diff
@@ -0,0 +1,244 @@
+            }
+        }
+
+        @Override
+        public void cancelAccumulation() {
+            final Cancellable cancellable = timerCancellableUpdater.getAndSet(this, EMPTY_SUBSCRIPTION_NO_THROW);
+            if (cancellable != null) {
+                cancellable.cancel();
+            }
+        }
+
+        private static int getSize(long stateSize) {
+            return (int) stateSize;
+        }
+
+        private static int getState(long stateSize) {
+            return (int) (stateSize >>> 32);
+        }
+
+        private static long buildStateSize(int state, int size) {
+            return (((long) state) << 32) | size;
+        }
+
+        private void schedulerTimer(long nanos) {
+            for (;;) {
+                final Cancellable currentCancellable = timerCancellable;
+                if (currentCancellable == EMPTY_SUBSCRIPTION_NO_THROW) {
+                    break;
+                } else {
+                    final Cancellable nextCancellable = executor.schedule(this::expireSignals, nanos, NANOSECONDS);
+                    if (timerCancellableUpdater.compareAndSet(this, currentCancellable, nextCancellable)) {
+                        // Current logic only has 1 timer outstanding at any give time so cancellation of
+                        // the current cancellable shouldn't be necessary but do it for completeness.
+                        if (currentCancellable != null) {
+                            currentCancellable.cancel();
+                        }
+                        break;
+                    } else {
+                        nextCancellable.cancel();
+                    }
+                }
+            }
+        }
+
+        private void expireSignals() {
+            final long nanoTime = executor.currentTime(NANOSECONDS);
+            TimeStampSignal<T> item;
+            for (;;) {
+                // read stateSize before peek, so if we poll from the queue we are sure to see the correct
+                // state relative to items in the queue.
+                final long currentStateSize = stateSize;
+                item = items.peek();
+                if (item == null) {
+                    break;
+                } else if (nanoTime - item.timeStamp >= ttlNanos) {
+                    final int currentSize = getSize(currentStateSize);
+                    if (stateSizeUpdater.compareAndSet(this, currentStateSize,
+                            buildStateSize(getState(currentStateSize) + 1, currentSize - 1))) {
+                        // When we add: we add to the queue we add first, then CAS sizeState.
+                        // When we remove: we CAS the atomic state first, then poll.
+                        // This avoids removing a non-expired item because if the "add" thread is running faster and
+                        // already polled "item" the CAS will fail, and we will try again on the next loop iteration.
+                        items.poll();
```

Comment:
I think we may still have a form of race between `accumulate(T)` and `expireSignals()` that can make otherwise good elements disappear.
```
items: [expired=true], maxItems = 1.
Thread1: accumulate. items.add(..); currentSize == maxItems, stateSizeUpdater.cas(..) == true; pause before items.poll();
items: [expired=true, expired=false]
Thread2: timer fires, expireSignals(); item != null && item is expired, stateSizeUpdater.cas(..) == true; items.poll();
items: [expired=false]
Thread1: items.poll(); break;
items: []
```

Reply:
you are correct, great catch! will fix.